### ISSUE TITLE
Can a non-included copy of jemalloc be used somehow?

### ISSUE THREAD

**Issue by xnox** [memkind/memkind#5]
In Ubuntu/Debian we generally avoid using embedded copies of code. It seems to me that jemalloc changes are small. Would we be able to use a shared copy of jemalloc at all? Would we need to propose some patches to debian maintainer of jemalloc to achieve this? E.g. if patches x,y,z are included in  https://tracker.debian.org/pkg/jemalloc we can just build depend on that in debian.

Doing above would be preferred to get this packaged in debian / ubuntu quickly.

**Comment by krzycz** [comment]
In fact, the changes you can see on "memkind" branch are relatively small, but there are also some other patches on "memkind-dev" branch that are to be integrated soon. These include thread cache enabling for extended arenas and the support for file-backed heaps.  I believe these changes should be transparent for non-*memkind* users of *jemalloc,* as the mentioned features are enabled only for extended arenas created by *memkind.*

The second issue is that *memkind* is built on *jemalloc 3.5.1*, not on the version 3.6.0. But this seems to be not a problem, as the delta between 3.5.1 and 3.6.0 is quite small, so we could probably sync to 3.6.0 easily.

**Comment by cmcantalupo** [comment]
Krzysztof has created a patch set for jemalloc version 3.6.0 to enable the memkind requirements here:

https://github.com/jemalloc/jemalloc/compare/master...krzycz:memkind-dev-3.6.0?expand=1

We will have to provide a configure option for memkind to enable build time selection for jemalloc as an internal statically linked library, or an external package that is linked dynamically (Fedora is using the static option).  We have supported both use cases in the past, and simply need to enable selection at configure time.  

One concern I have is about the --with-jemalloc-prefix configure option for jemalloc.  It seems that the Ubuntu package does not provide this configure option, and so the jemalloc library exports the ISO C malloc() family of APIs.  We will have to provide some symmetric configuration option for memkind to enable jemalloc prefix selection (or omission).  The memkind dev branch currently has a hard-coded prefix of "jemk_", and in our last tagged release (memkind 0.2.2) we used "je_".
